modules/navier_stokes/src/kernels/INSMomentumSUPG.C

```diff
@@ -0,0 +1,326 @@
+/****************************************************************/
+
+#include "INSMomentumSUPG.h"
+
+template <>
+InputParameters
+validParams<INSMomentumSUPG>()
+{
+  InputParameters params = validParams<Kernel>();
+
+  // Coupled variables
+  params.addRequiredCoupledVar("u", "x-velocity");
+  params.addCoupledVar("v", 0, "y-velocity"); // only required in 2D and 3D
+  params.addCoupledVar("w", 0, "z-velocity"); // only required in 3D
+  params.addRequiredCoupledVar("p", "pressure");
+
+  // Required parameters
+  params.addRequiredParam<RealVectorValue>("gravity", "Direction of the gravity vector");
+  params.addRequiredParam<unsigned>(
+      "component",
+      "0,1,2 depending on if we are solving the x,y,z component of the momentum equation");
+
+  // Optional parameters
+  params.addParam<bool>("integrate_p_by_parts",
+                        true,
+                        "Allows simulations to be run with pressure BC if set to false");
+  params.addParam<MaterialPropertyName>("mu_name", "mu", "The name of the dynamic viscosity");
+  params.addParam<MaterialPropertyName>("rho_name", "rho", "The name of the density");
+
+  return params;
+}
+
+INSMomentumSUPG::INSMomentumSUPG(const InputParameters & parameters)
+  : Kernel(parameters),
+
+    // Coupled variables
+    _u_vel(coupledValue("u")),
+    _v_vel(coupledValue("v")),
+    _w_vel(coupledValue("w")),
+    _p(coupledValue("p")),
+
+    // Gradients
+    _grad_u_vel(coupledGradient("u")),
+    _grad_v_vel(coupledGradient("v")),
+    _grad_w_vel(coupledGradient("w")),
+    _grad_p(coupledGradient("p")),
+
+    // Variable numberings
+    _u_vel_var_number(coupled("u")),
+    _v_vel_var_number(coupled("v")),
+    _w_vel_var_number(coupled("w")),
+    _p_var_number(coupled("p")),
+
+    // parameters
+    _gravity(getParam<RealVectorValue>("gravity")),
+    _component(getParam<unsigned>("component")),
+    _integrate_p_by_parts(getParam<bool>("integrate_p_by_parts")),
+
+    // Material properties
+    _mu(getMaterialProperty<Real>("mu_name")),
+    _rho(getMaterialProperty<Real>("rho_name"))
+{
+}
+
+Real
+INSMomentumSUPG::computeQpResidual()
+{
+  RealVectorValue U(_u_vel[_qp], _v_vel[_qp], _w_vel[_qp]);
+  Real alpha;
+  if (_mu[_qp] < std::numeric_limits<double>::epsilon())
+    alpha = 1;
+  else if (U.norm() < std::numeric_limits<double>::epsilon())
+  {
+    alpha = 0;
+    return 0;
+  }
+  else
+  {
+    Real grid_Pe = _current_elem->hmax() * U.norm() / (2. * _mu[_qp]);
+    alpha = 1. / std::tanh(grid_Pe) - 1. / grid_Pe;
+  }
+  Real PG_test = alpha * _current_elem->hmax() / 2. * U / U.norm() * _grad_test[_i][_qp];
+
+  Real convective_part = _rho[_qp] * U * _grad_u[_qp] * PG_test;
+
+  // Note that integrating pressure by parts would imply applying a second
+  // derivative to the test function, which we are not equipped to do (?)
+  Real pressure_part = 0;
+  if (!_integrate_p_by_parts)
+    pressure_part = _grad_p[_qp](_component) * PG_test;
```

Comment:
Per our discussion, I think the stabilization terms should involve grad(p) regardless of whether or not we integrate by parts.